packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/internal/BaseSubscriptionSetImpl.kt

```diff
@@ -0,0 +1,97 @@
+package io.realm.mongodb.internal
+
+import io.realm.BaseRealm
+import io.realm.RealmObject
+import io.realm.internal.BaseRealmImpl
+import io.realm.internal.interop.RealmBaseSubscriptionSetPointer
+import io.realm.internal.interop.RealmInterop
+import io.realm.internal.interop.RealmSubscriptionPointer
+import io.realm.internal.query.ObjectQuery
+import io.realm.mongodb.sync.BaseSubscriptionSet
+import io.realm.mongodb.sync.Subscription
+import io.realm.mongodb.sync.SubscriptionSetState
+import io.realm.query.RealmQuery
+import kotlinx.atomicfu.AtomicRef
+
+internal abstract class BaseSubscriptionSetImpl<T : BaseRealm>(
+    protected val realm: T,
+) : BaseSubscriptionSet {
+
+    protected abstract val nativePointer: AtomicRef<out RealmBaseSubscriptionSetPointer>
+    protected abstract fun getIteratorSafePointer(): RealmBaseSubscriptionSetPointer
+
+    protected fun checkClosed() {
+        (realm as BaseRealmImpl).realmReference.checkClosed()
+    }
+
+    override fun <T : RealmObject> findByQuery(query: RealmQuery<T>): Subscription? {
+        val queryPointer = (query as ObjectQuery).queryPointer
+        val subscriptionPointer: RealmSubscriptionPointer? = RealmInterop.realm_sync_find_subscription_by_query(
```

Comment:
Doesn't look safe if the native pointer can be updated. Either we need to work on the same pointer throughout the method, lock for updates or document that we are not thread safe.